In [13]:
import numpy as np
import pandas as pd
heroes = pd.read_csv("heroes_information.csv")

In [14]:
###instalar desde anaconda navigator######
from pandasql import *

In [15]:
pysqldf = lambda q: sqldf(q,globals())

In [16]:
query = " SELECT * FROM heroes limit 10"

In [17]:
pysqldf(query)

id           name  Gender Eye color               Race Hair color  Height  \
0   0         A-Bomb    Male    yellow              Human    No Hair   203.0   
1   1     Abe Sapien    Male      blue      Icthyo Sapien    No Hair   191.0   
2   2       Abin Sur    Male      blue            Ungaran    No Hair   185.0   
3   3    Abomination    Male     green  Human / Radiation    No Hair   203.0   
4   4        Abraxas    Male      blue      Cosmic Entity      Black   -99.0   
5   5  Absorbing Man    Male      blue              Human    No Hair   193.0   
6   6    Adam Monroe    Male      blue               None      Blond   -99.0   
7   7   Adam Strange    Male      blue              Human      Blond   185.0   
8   8       Agent 13  Female      blue               None      Blond   173.0   
9   9      Agent Bob    Male     brown              Human      Brown   178.0   

           Publisher Skin color Alignment  Weight  
0      Marvel Comics       None      good   441.0  
1  Dark Horse Comics       blue      good    65.0  
2          DC Comics        red      good    90.0  
3      Marvel Comics       None       bad   441.0  
4      Marvel Comics       None       bad   -99.0  
5      Marvel Comics       None       bad   122.0  
6       NBC - Heroes       None      good   -99.0  
7          DC Comics       None      good    88.0  
8      Marvel Comics       None      good    61.0  
9      Marvel Comics       None      good    81.0

**1. Cuál es el nombre del primer superhéroe de la tabla heroes?**

In [18]:
query1 = " SELECT * FROM heroes limit 1"
pysqldf(query1)

id    name Gender Eye color   Race Hair color  Height      Publisher  \
0   0  A-Bomb   Male    yellow  Human    No Hair   203.0  Marvel Comics   

  Skin color Alignment  Weight  
0       None      good   441.0

*Respuesta: A-Bomb*

**2. Cuántas casas publicadoras existen?**

In [21]:
query2 = " SELECT count(DISTINCT Publisher) FROM heroes"
pysqldf(query2)

count(DISTINCT Publisher)
0                         24

**3. Cuántos superhéroes miden más de 2 metros (200 cms)**

In [22]:
query3 = " SELECT count(*) FROM heroes WHERE Height > 200"
pysqldf(query3)

count(*)
0        59

**4. Cuántos superhéroes son humanos y miden más de 2 metros (200 cms)?**

In [23]:
query4 = " SELECT count(*) FROM heroes WHERE Height > 200 AND Race = 'Human'"
pysqldf(query4)

count(*)
0        12

**5. Cuántos superhéroes pesan más de 100 lbs y menos de 200 lbs?**

In [24]:
query5 = " SELECT count(*) FROM heroes WHERE Weight BETWEEN 101 AND 199"
pysqldf(query5)

count(*)
0        98

**6. Cuántos superhéroes tienen los ojos de color azul o rojo?**

In [27]:
query6 = " SELECT count(*) FROM heroes WHERE [Eye color] IN ('red','blue')"
pysqldf(query6)

count(*)
0       271

**7. Cuántos superhéroes son Human, Mutant y tienen el pelo color Verde o son Vampiros con pelo Negro**

In [40]:
query7 = " SELECT count(*) FROM heroes WHERE (Race IN ('Human','Mutant') AND [Hair color]='Green') OR (Race='Vampire' AND [Hair color]='Black')"
pysqldf(query7)

count(*)
0         5

**8. Cuál es el nombre del primer superhéroe si ordeno la tabla por raza en orden descendente?**

In [42]:
query8 = " SELECT name FROM heroes ORDER BY Race DESC limit 1"
pysqldf(query8)

name
0  Solomon Grundy

**9. Cuántos superhéroes son de género masculino y cuántas de género femenino?**

In [46]:
query9 = " SELECT count(*) FROM heroes GROUP BY Gender"
pysqldf(query9)

count(*)
0        29
1       200
2       505

*Respuesta: 200 mujeres y 505 hombres*

**10. Cuántas casas publicadoras tienen más de 15 superhéroes?**

In [57]:
query10 = " SELECT count(id) FROM heroes GROUP BY Publisher HAVING COUNT(id)>15 "
pysqldf(query10)

count(id)
0        215
1         18
2        388
3         19

*Respuesta: 4 casas publicadoras*